### DISCUSSION TITLE
`RRule` vs `RRuleSet`

### DISCUSSION THREAD

**Discussion by omid** [fmeringdal/rust-rrule#55] [General]
Hello,

What are the differences between `RRuleSet` and `RRule`?
Can we merge them? (or always use `RRuleSet` if it's a superset for `RRule`)
What are the equivalent RFC standard of them? ([this](https://datatracker.ietf.org/doc/html/rfc5545)?)

**Comment by fmeringdal** [comment]
Copying my answer from #45:

I think we should keep both types. These are some of the reasons why I think so:

- This library is largely a port of the javascript and python rrule implementation, they separate these types.
- I imagine that if we removed RRule, then we will need to complicate the API even more if the user wants to disable EXDATES, EXRULES and RDATES. I also think that some users not familiar with the spec might be surprised by the fact that the string parsing will allow these "exotic variants" and lead to unexpected behavior if they accept strings from end users.

**Comment by omid** [comment]
> This library is largely a port of the javascript and python rrule implementation, they separate these types.

I think it's not a good reason to have an API. Even if so, they may have a bad structure. I mean, we should know the exact structure and reason behind it.

> I imagine that if we removed RRule, then we will need to complicate the API even more if the user wants to disable EXDATES, EXRULES and RDATES. I also think that some users not familiar with the spec might be surprised by the fact that the string parsing will allow these "exotic variants" and lead to unexpected behavior if they accept strings from end users.

Having exact same logics in different places in the source code is wrong. Imagine an end user sends a VEVENT string with an EXDATE and the developer used rrule, then the result will be absolutely something wrong!

Now `RRule` is correct! But `RRuleSet` is wrong since it doesn't have TZ. And the `Display` and `FromStr` doesn't work fine for it, because it doesn't follow the standard.

**Comment by fmeringdal** [comment]
This is a good discussion to have and definitely needs to be clarified before going to v1. This is quite a significant change so we should really think through the pros and cons here. 

> I think it's not a good reason to have an API. Even if so, they may have a bad structure. I mean, we should know the exact structure and reason behind it.

I agree that we should know the exacts reasons for it. Looking at the other `rrule` implementations for the other languages like js, python, php and go they do separate between `RRule` and `RRuleSet`. The swift and ruby libraries do only have one type, but they do not support `EXRULE` so then it makes even more sense. My point is that I think there are reasons that they have separated those types and we should try to figure out what they are. It could also be the case that everyone has just blindly copied the python implementation of RRule (like I did initially) and never gave any thought to whether they should merge `RRule` and `RRuleSet` or not.

> Having exact same logics in different places in the source code is wrong. Imagine an end user sends a VEVENT string with an EXDATE and the developer used rrule, then the result will be absolutely something wrong!

I see your point. But I think just a quick look at the documentation will make the user understand that `RRule` is for a single `RRULE` line whereas `RRuleSet` is for composing `RRULE`, `EXRULE` etc. I don't think there is that much duplication of logic in the source code expect that the `RRule` and `RRuleSet` exposes the same api?

> Now RRule is correct! But RRuleSet is wrong since it doesn't have TZ. And the Display and FromStr doesn't work fine for it, because it doesn't follow the standard.

This is a good point. Currently the start dates and timezones are defined in the individual recurrence rules it contains (i.e. EXRULES and RRULES), but that is problematic for the `Display` implementation as there might be multiple timezones and start dates to select as `DTSTART`. `FromStr` should work if all nested recurrence rules uses the start date and timezone specified in the string? Adding dtstart + tzid  fields to `RRuleSet` and forcing the nested recurrence rules to use those start dates could be a possible solution.

**Comment by omid** [comment]
> This is a good discussion to have and definitely needs to be clarified before going to v1.

I think we shouldn't block releasing v1.0 to this discussion. We can release v2.0 one day :)

> The swift and ruby libraries do only have one type, but they do not support EXRULE so then it makes even more sense.

I think the main reason for the differences between the implementations is that how strict they want to be. Based on the RFC, `EXRULE` is obsolete and in some libraries, they drop it. In the other hand, again based on RFC, there SHOULD be one `RRULE`, and more than that is not illegal, but is an undefined behavior!

But I agree with what we have now :) I mean I agree we need to have these 3 structs (maybe with a little different structure).

The problem I see is that we shouldn't have `dt_start` and `tz` in the `RRule`. I know now we need them for validation and then conversion to iterator, and this is what I'm for to change.

On the other hand, we should have `dt_start` and `tz` in `RRuleSet`. And send it around whenever they are needed.

> I don't think there is that much duplication of logic in the source code expect that the RRule and RRuleSet exposes the same api?

You are right, the main logics are well separated. Do we have a validator for `RRuleSet`?

>  FromStr should work if all nested recurrence rules uses the start date and timezone specified in the string?

AFAICS, based on RFC, `RRule` (and `EXRULE`) doesn't have any timezone and also start_dt, and they are always UTC or local. 

> Adding dtstart + tzid fields to RRuleSet and forcing the nested recurrence rules to use those start dates could be a possible solution.

And this is what I likeeeee + if we don't have it yet, a validation on RRuleSet. I don't know if we should create a set like `RRule` and `RRuleProperties` or not.

PS: One another point regarding my recent changes, after we did the start_dt and tz separation mentioned above, `Display` impl of `RRule` and `RRuleProperties` should be exactly the same! Actually, `RRule` can just call the `RRuleProperties` impl! And we couldn't have `FromStr` for `RRule` because we cannot validate it.

**Comment by fmeringdal** [comment]
If we removed start date and timezone fields from `RRule` then `RRuleProperties` would contain exactly the same and only difference is that properties of `RRule` are "validated"? That would also mean that `RRule` does not implement `IntoIterator` and only way to generate occurences is via `RRuleSet` which contains the start date and timezone? I am not so sure we want `RRule` to be a direct mapping to the `RRULE` line in the RFC which as you say does not have a start date and timezone. 

Anyways, to summarise I think we both agree on this change:
- `RRuleSet` should have a start date and timezone field which should override any nested start date and timezone fields (until we potentially remove start date and timezone from `RRule` at a later stage).

**Comment by omid** [comment]
> If we removed start date and timezone fields from RRule then RRuleProperties would contain exactly the same and only difference is that properties of RRule are "validated"?

Yes, we can follow this pretty state machine, which I tried once, but exactly because of start date and timezone it was not so pretty :) https://hoverbear.org/blog/rust-state-machine-pattern/

> That would also mean that RRule does not implement IntoIterator and only way to generate occurences is via RRuleSet which contains the start date and timezone?

Yes and no. `RRule` can have a custom `IntoIterator`, actually we can convert `RRule` to the existing `RRuleIter` struct, by a custom method, like `into_iter_with_ctx` or a custom impl of `into_iter` (but with start date and tz as input param).

> I am not so sure we want RRule to be a direct mapping to the RRULE line in the RFC which as you say does not have a start date and timezone.

What are the concerns?

> RRuleSet should have a start date and timezone field which should override any nested start date and timezone fields (until we potentially remove start date and timezone from RRule at a later stage).

Yes, I agree.

If you ask me to have a very short list of the tasks needed to be done (from what I see now)

- Add start date and tz to `RRuleSet`
- Remove start date and tz from `RRule`
- Remove `IntoIterator` of `RRule` and change it to a custom (private) method which accepts start date and tz as input param.
- Add a validator for `RRuleSet` and validate the whole structure and add another state to that (validated/not-yet-validated)
- (bonus) Merge different validated and not-yet-validated states of `RRuleSet` and `RRule` with the _pretty_ state machine.

**Comment by fmeringdal** [comment]
> Yes, we can follow this pretty state machine.

I was thinking that something simpler like this could work as well by just leveraging generics:
```rust
struct Unvalidated;
struct Validated;

struct RRuleProperties<Stage = Unvalidated> {
  freq: Frequency,
  ...
  stage: std::marker::PhantomData<Stage>,
}

impl RRuleProperties<Stage = Unvalidated> {
  fn validate(self) -> Result<Self<Validated>, ValidatonError> {
    ....
  }
}

struct RRule {
  dt_start: Datetime,
  properties: RRuleProperties<Validated>
}
```

Then we do not have to define so boilerplate "machinery" code, but we can ofc discuss impl details further when we have created the issue/PR.

> What are the concerns?

My concern is that I feel that it would make the API more awkward if we have to call `into_iter_with_ctx ` or somethine like that on `RRule` to get an iterator. Somehow I really like the current abstractions.

> Add start date and tz to RRuleSet

Yes!

> Add a validator for RRuleSet and validate the whole structure and add another state to that (validated/not-yet-validated)

Let's specify what exactly needs to be validated for an `RRuleSet` when we create tickets for it. Most should already be validated if it only contains validated `RRuleProperties` / `RRule`.  It could also be the case that we can make the API for `RRuleSet` such that it is never possible to construct an invalid `RRuleSet` in the first place.

> Remove start date and tz from RRule

Not yet, more thinking / discussions needed.

> Remove IntoIterator of RRule and change it to a custom (private) method which accepts start date and tz as input param.

Not yet, more thinking / discussions needed.

> (bonus) Merge different validated and not-yet-validated states of RRuleSet and RRule with the pretty state machine.

Not exactly sure what you mean here, but seems like something that can be described in a new issue when we get to a stage where we can implement this.

**Comment by omid** [comment]
> I was thinking that something simpler like this could work as well by just leveraging generics.

Your code is almost that _pretty_ state machine, and I agree with the code you wrote.

> My concern is that I feel that it would make the API more awkward if we have to call into_iter_with_ctx or somethine like that on RRule to get an iterator. Somehow I really like the current abstractions.

This `into_iter_with_ctx` is in the private API, nobody can ever be able to iterate over a simple `RRule`. They actually need `RRuleSet` and there is a public API there.

> Let's specify what exactly needs to be validated for an RRuleSet when we create tickets for it. Most should already be validated if it only contains validated RRuleProperties / RRule.

`RRuleSet` should have a list of validated `RRule`s. Users may not be able to validate RRule separately, we can even remove that from the public API.

`RRuleSet` can be something like this:
```
#[derive(Debug, Clone)]
pub struct RRuleSet<S> {
    pub rrule: Vec<RRule<S>>,
    pub rdate: Vec<DateTime>,
    pub exrule: Vec<RRule<S>>,
    pub exdate: Vec<DateTime>,
    pub dt_start: DateTime,
    pub tz: Tz,
}
```
So a `Validated` `RRuleSet` also contains `Validated` `RRule`s.

> Not exactly sure what you mean here, but seems like something that can be described in a new issue when we get to a stage where we can implement this.

I meant like the code sample you put here. So seems like we already kinda agreed on this :D

**Comment by fmeringdal** [comment]
I am experimenting a little bit with how the public API would look like if we removed start date from `RRule`. I think if possible we should hide the generic parameter from the public API. So that `RRuleSet` contains only `RRuleProperties<Validated>` values. Also since `RRule` and `RRuleProperties` will be exactly the same if we removed the start date then it makes sense to only use `RRuleProperties` and remove `RRule`.

```rust
struct RRuleProperties<S> {
  freq: Frequency,
  ...
}

impl RRuleProperties<Validated> {
  pub(crate) fn into_iter(self, dt_start: Datetime, tz: Tz) -> RRuleIter {}
}

struct RRuleSet {
  pub rrule: Vec<RRuleProperties<Validated>>,
  pub rdate: Vec<DateTime>,
  pub exrule: Vec<RRuleProperties<Validated>>,
  pub exdate: Vec<DateTime>,
  pub dt_start: DateTime,
  pub tz: Tz,
}

// Usage 1: Construction from properties
let mut set = RRuleSet::new(dtstart);
let props = RRuleProperties {
  freq: Daily,
  ..Default::default()
};
set.rrule(props).unwrap(); // RRuleSet::rule returns Result<(), ValidationError> 
set.all(100);

// Usage 2: Construction from string
let mut set: RRuleSet = "DTSTART:...".parse().unwrap(); // If parsing from string was successful, then we know the RRuleSet is valid.
set.all(100);

// Usage 3: Serde
#[derive(Deserialize, Serialize)]
struct RequestBody {
  rrule_set: RRuleSet // If deserializing was successful, then we know the RRuleSet is valid.
}

```

**Comment by omid** [comment]
I agree generally. I'm also experiencing some things. It's a huge change, everywhere :D
The only thing I suggest here is that, better to merge `RRule` and `RRuleProperties` into `RRule` :)

I think I also agree that `RRuleSet` is always validated.

**Comment by fmeringdal** [comment]
> The only thing I suggest here is that, better to merge RRule and RRuleProperties into RRule :)

Agreed.

@ralpha do you have any immediate concerns with not exposing `RRule` and only having `RRuleSet` as the public type?

I think my only concerns are these:
1. `RRuleSet` enables more features with `RDATE`, `EXDATE` and `EXRULE` which the user might not want to use. Should we have an option for them to opt out of that? Or maybe make the `RRule::into_iter` method public to allow for such things.
2. Other libraries that implements the RRULE RFC and supports `EXRULE` exposes both an `RRule` and `RRuleSet`. That does not necessarily mean that it is the optimal approach, but it could be that they have use-cases in mind which we have not been thinking about here.

**Comment by omid** [comment]
> do you have any immediate concerns with not exposing RRule and only having RRuleSet as the public type?

We still can make `RRule` public. But we don't need to public everything.
For example, that doesn't need to have a public `RRuleIter`, so no iteration is possible on `RRule`. But the existing `build` method in `RRule` can stay public. We may need a public `validate` also, to convert `Unvalidated` to `Validated`.

I'm saying, since we have it, let's make it public for somebody who required that (including myself, maybe :D).

> RRuleSet enables more features with RDATE, EXDATE and EXRULE which the user might not want to use. Should we have an option for them to opt out of that? Or maybe make the RRule::into_iter method public to allow for such things.

I think I disagree to have an option to disable them. Mostly people will generate `RRuleSet` from a string and they are free to use or not use the public API.
